### PR TITLE
feat: Add support for completing `cargo update <TAB>`

### PR DESCRIPTION
**Pull request by shannmu**
### What does this PR try to resolve?
Tracking issue https://github.com/rust-lang/cargo/issues/14520

Add custom completer for `cargo update <TAB>`

### CODE REVIEW

File: src/cargo/util/command_prelude.rs [hunk 1]
```diff
@@ -1067,6 +1068,29 @@ fn get_targets_from_metadata() -> CargoResult<Vec<Target>> {
     Ok(targets)
 }
 
+pub fn get_dependency_name_candidates() -> Vec<clap_complete::CompletionCandidate> {
+    get_dependencies_from_metadata()
+        .unwrap_or_default()
+        .into_iter()
+        .map(|dep| clap_complete::CompletionCandidate::new(dep.name_in_toml().as_str()))
```
Comment (epage) [inline src/cargo/util/command_prelude.rs:1075]: Let's include the descriptions

Comment (epage) [inline src/cargo/util/command_prelude.rs:1075]: I think this is supposed to be package name, not dependency name (`name_in_toml`)

Comment (epage) [inline src/cargo/util/command_prelude.rs:1075]: package names may not always be sufficient.

In clap, run
```console
$ cargo update clap
error: There are multiple `clap` packages in your project, and the specification `clap` is ambiguous.
Please re-run this command with one of the following specifications:
  clap@4.5.9
  clap@4.5.17
```

`PackageIdSpec::query::minimize` has a crude algorithm for uniquifing things.  We should have a more robust form.
- Put all packages into a map of `name -> Vec<Package>`
- For any entries with duplicates but unique versions, report `name@version`
- If there are duplicate versions, report the full `PackageIdSpec`

File: src/cargo/util/command_prelude.rs [hunk 2]
```diff
@@ -1067,6 +1068,29 @@ fn get_targets_from_metadata() -> CargoResult<Vec<Target>> {
     Ok(targets)
 }
 
+pub fn get_dependency_name_candidates() -> Vec<clap_complete::CompletionCandidate> {
```
Comment (epage) [inline src/cargo/util/command_prelude.rs:1071]:
```suggestion
pub fn get_package_candidates() -> Vec<clap_complete::CompletionCandidate> {
```

File: src/cargo/util/command_prelude.rs [hunk 3]
```diff
@@ -1067,6 +1068,29 @@ fn get_targets_from_metadata() -> CargoResult<Vec<Target>> {
...
+    get_dependencies_from_metadata()
+        .unwrap_or_default()
+        .into_iter()
+        .map(|dep| clap_complete::CompletionCandidate::new(dep.name_in_toml().as_str()))
+        .collect::<Vec<_>>()
+}
+
+fn get_dependencies_from_metadata() -> CargoResult<Vec<Dependency>> {
```
Comment (epage) [inline src/cargo/util/command_prelude.rs:1079]:
```suggestion
fn get_packages() -> CargoResult<Vec<Dependency>> {
```

File: src/cargo/util/command_prelude.rs [hunk 4]
```diff
@@ -1067,6 +1068,29 @@ fn get_targets_from_metadata() -> CargoResult<Vec<Target>> {
...
+
+    let dependencies = ws
+        .members()
+        .flat_map(|pkg| pkg.dependencies().into_iter().cloned())
+        .collect::<HashSet<_>>()
+        .into_iter()
+        .collect::<Vec<_>>();
+
+    Ok(dependencies)
+}
```
[the remark this answers is not in the document]
Comment (shannmu) [reply src/cargo/util/command_prelude.rs:1092]: > We can either updated our `GlobalContext` to be quiet (should probably go back and do that to all of these completions anyways)

Are you referring to updating the `GlobalContext`(in the main function) within the custom completer function?

If this is what you mean, I think we will face a problem: the custom completer function cannot carry additional parameters. Capturing `&mut gctx` with a closure would likely be restricted by lifetime limitations. If there are other methods I haven’t considered, please let me know. Thank you!

File: src/cargo/util/command_prelude.rs (hunk 1 above)
Comment (shannmu) [reply src/cargo/util/command_prelude.rs:1075]: Thanks for your guidance!

File: src/cargo/util/command_prelude.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (epage) [reply src/cargo/util/command_prelude.rs:1092]: > `    let gctx = GlobalContext::new(shell::Shell::new(), cwd.clone(), cargo_home_with_cwd(&cwd)?);`

For the `gctx` we have, we should force it to be quiet.  The primary-path in the CLI handles that through calling [`configure`](https://docs.rs/cargo/latest/cargo/util/context/struct.GlobalContext.html#method.configure).  Unsure if there are more direct ways of configuring it.  We might also want to call `configure` anyways as our context is only partially initialized and I've not audited it for any problems we can run into without fully initializing it.

File: src/cargo/util/command_prelude.rs [hunk 5]
```diff
@@ -1174,6 +1178,122 @@ fn get_target_triples_from_rustc() -> CargoResult<Vec<clap_complete::CompletionC
         .collect())
 }
 
+pub fn get_package_candidates() -> Vec<clap_complete::CompletionCandidate> {
```
Comment (epage) [inline src/cargo/util/command_prelude.rs:1181]: we should clarify this is `get_pkg_id_spec_candidates` with this change.

I only realized this as I wondered if we could reuse this for `cargo info` (maybe) and `cargo add` (no)

File: src/cargo/util/command_prelude.rs [hunk 6]
```diff
@@ -1174,6 +1178,122 @@ fn get_target_triples_from_rustc() -> CargoResult<Vec<clap_complete::CompletionC
...
+                })
+                .collect::<Vec<_>>()
+        })
+        .collect::<Vec<_>>()
+}
+
+fn get_packages() -> CargoResult<Vec<Package>> {
+    let cwd = std::env::current_dir()?;
+    let mut gctx = GlobalContext::new(shell::Shell::new(), cwd.clone(), cargo_home_with_cwd(&cwd)?);
+    gctx.configure(0, true, None, false, true, false, &None, &[], &[])?;
```
Comment (epage) [inline src/cargo/util/command_prelude.rs:1266]: Could you split these out into variables to clarify their names?

We'll likely need this in all of the completions that create a context, so maybe we should provide a helper for this?  Feels too specialized and accident prone to live on `GlobalContext` but unsure.

File: src/cargo/util/command_prelude.rs [hunk 7]
```diff
@@ -1174,6 +1178,122 @@ fn get_target_triples_from_rustc() -> CargoResult<Vec<clap_complete::CompletionC
...
+                                    .description
+                                    .to_owned()
+                                    .map(From::from),
+                            )
+                        })
+                        .collect::<Vec<_>>()
+                })
+                .collect::<Vec<_>>()
+        })
+        .collect::<Vec<_>>()
```
Comment (epage) [inline src/cargo/util/command_prelude.rs:1260]: 1. are all of these `collect`s needed?  I thought `flat_map` accepted an `IntoIterator`
2. This feels like too complex for combinators for me

File: src/cargo/util/command_prelude.rs (hunk 5 above)
Comment (shannmu) [reply src/cargo/util/command_prelude.rs]: Change done! Thanks!

File: src/cargo/util/command_prelude.rs (hunk 6 above)
Comment (shannmu) [reply src/cargo/util/command_prelude.rs]: > We'll likely need this in all of the completions that create a context, so maybe we should provide a helper for this?

I will refactor other completers used `gctx` in another pr.

> Feels too specialized and accident prone to live on GlobalContext but unsure.

For now, I make this function a non-pub function for now. I think putting it in the impl of `GlobalContext` is too specialized.

File: src/cargo/util/command_prelude.rs (hunk 7 above)
Comment (shannmu) [reply src/cargo/util/command_prelude.rs]: Modify it a clearer form! Thanks!

**Comment by shannmu** [comment]
The completion speed of this feature is not very ideal. On my local machine, I can feel a slight lag.

**Comment by epage** [comment]
> The completion speed of this feature is not very ideal. On my local machine, I can feel a slight lag.

Looked at it using https://doc.crates.io/contrib/tests/profiling.html

image

Note: this was done on a debug build, doing `cargo update [TAB]` within Cargo.  There was a pause but not too terrible.

- We should probably switch approaches so tracing output is flushed
- We have some unidentified gaps in time
- I've been looking to improve package load times which is about half of the total execution time
- If we're okay with dropping the package description, we could remove the `download_accessible` call by switching from `resolve_ws_with_opts` to `resolve_ws`, removing all of the manifest parsing time
  - If we want to keep descriptions, I suspect there is a faster way than running the feature resolver (which `resolve_ws_with_opts` does)
- The resolve times I've already been having conversations on how to speed up

**Comment by shannmu** [comment]
> The resolve times I've already been having conversations on how to speed up

May I ask where is the discussion about how to speed up resolve times?

**Pull request merged**